[SPMD] Introduce remove_copy_for_optimizer optimization

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98580
* #98579
* #98578
* #98285
* #98284
* #98182

This PR adds the ability to remove unused `copy_` (`len(node.users) == 0`) that generated by tracing the optimizer.

Differential Revision: [D44761556](https://our.internmc.facebook.com/intern/diff/D44761556/)

fegin committed:
[SPMD] Introduce remove_copy_for_optimizer optimization

This PR adds the ability to remove unused `copy_` (`len(node.users) == 0`) that generated by tracing the optimizer.

Differential Revision: [D44761556](https://our.internmc.facebook.com/intern/diff/D44761556/)

[ghstack-poisoned]

fegin commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -481,3 +482,44 @@ def schedule_comm_wait(gm: IterGraphModule) -> None:
+    remove_candidates: Set[fx.Node] = set()
+    for node in reversed(gm.graph.nodes):
+        if node.users:
+            continue
```
> @mrshenli's internal review: curious, this is the same as DCE? but specialized for foreach? If so, will it work if we just do DCE before going through sparse expansion? cc., @yifuwang 

Not the same as DCE. `fx.Graph` marks `copy_` as side effect node. So a `copy_` won't be removed even if it has no users. That's why we are seeing the orphan gradients `copy_`. However, this pass exists for if and only if the graph has other `copy_` which has zero users and we want to preserve them. Otherwise, we can simply the pass by simply traverse all the `copy_` nodes and remove those which have zero users. And I also think it makes sense to do this after the expansion. I only need this pass to continue the optimizer work at this moment.

mrshenli replied:
> fx.Graph marks copy_ as side effect node. 

Oh I see, this is new to me

fegin commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -481,3 +482,44 @@ def schedule_comm_wait(gm: IterGraphModule) -> None:
+            if visiting.op == OP.CALL_FUNCTION and str(visiting.target).startswith(
+                ("aten._foreach_", "aten._fused_")
+            ):
+                should_remove = True
```
> @mrshenli's internal review: any reason that we need to continue BFS backward the graph instead of stop after seeing foreach?

It is safer to traverse all the same distances and put the nodes into the candidates. It won't hurt us because the removing logic will check if `len(users) == 0` -- we won't remove those nodes that are still being used.

fegin commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -481,3 +482,44 @@ def schedule_comm_wait(gm: IterGraphModule) -> None:
+                if isinstance(parent, fx.Node):
+                    nodes.append(parent)
+        if should_remove:
+            remove_candidates.update(copy_ancestors)
```
> @mrshenli's internal review: what if some of the ancestors have other consumers?

Same reason as above, we won't remove those ancestors if they still have users.

fegin committed:
Update on "[SPMD] Introduce remove_copy_for_optimizer optimization"


This PR adds the ability to remove unused `copy_` (`len(node.users) == 0`) that generated by tracing the optimizer.

Differential Revision: [D44761556](https://our.internmc.facebook.com/intern/diff/D44761556/)

[ghstack-poisoned]

mrshenli commented on torch/distributed/_spmd/graph_optimization.py:
```diff
@@ -476,3 +476,44 @@ def schedule_comm_wait(gm: IterGraphModule) -> None:
+
+    for node in reversed(gm.graph.nodes):
+        if node.users:
+            continue
```
Oh I see, this is where you skip ancestors with other users

fegin committed:
Update on "[SPMD] Introduce remove_copy_for_optimizer optimization"


This PR adds the ability to remove unused `copy_` (`len(node.users) == 0`) that generated by tracing the optimizer.

Differential Revision: [D44761556](https://our.internmc.facebook.com/intern/diff/D44761556/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce remove_copy_for_optimizer optimization"


This PR adds the ability to remove unused `copy_` (`len(node.users) == 0`) that generated by tracing the optimizer.

Differential Revision: [D44761556](https://our.internmc.facebook.com/intern/diff/D44761556/)

[ghstack-poisoned]

fegin committed:
Update on "[SPMD] Introduce remove_copy_for_optimizer optimization"


This PR adds the ability to remove unused `copy_` (`len(node.users) == 0`) that generated by tracing the optimizer.

Differential Revision: [D44761556](https://our.internmc.facebook.com/intern/diff/D44761556/)

[ghstack-poisoned]

mrshenli commented:
@pytorchbot merge

mrshenli commented:
@pytorchbot merge -f "Test failure is irrelevant"
